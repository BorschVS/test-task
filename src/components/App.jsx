import { lazy } from 'react';
import { Route, Routes } from 'react-router-dom';

import { Layout } from './Layout';

const Home = lazy(() => import('../pages/Home'));
const Flights = lazy(() => import('../pages/Flights'));
const Hotels = lazy(() => import('../pages/Hotels'));

const App = () => (
  <Routes>
    <Route path="/" element={<Layout />}>
      <Route index element={<Home />} />
      <Route path="flights" element={<Flights />} />
      <Route path="hotels" element={<Hotels />} />
    </Route>
  </Routes>
);

export default App;
