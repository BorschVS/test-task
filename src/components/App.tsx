import { FC, lazy } from 'react';
import { Route, Routes } from 'react-router-dom';

import { Layout } from 'components';

import { Page } from 'types/interfaces';

const Home = lazy(() => import('../pages/Home'));
const Flights = lazy(() => import('../pages/Flights'));
const Hotels = lazy(() => import('../pages/Hotels'));

const pages: Page[] = [
  { path: '/', element: <Home />, index: true },
  { path: 'flights', element: <Flights /> },
  { path: 'hotels', element: <Hotels /> },
];

const App: FC = () => (
  <Routes>
    <Route path="/" element={<Layout />}>
      {pages.map((page) =>
        page.index ? (
          <Route key={page.path} index element={page.element} />
        ) : (
          <Route key={page.path} path={page.path} element={page.element} />
        )
      )}
    </Route>
  </Routes>
);

export default App;
