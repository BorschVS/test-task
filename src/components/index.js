export { default as App } from './App';
export { default as CategoryFilter } from './CategoryFilter';
export { default as FlightCard } from './FlightCard';
export { default as TransferFilter } from './TransferFilter';
export { default as ResponsiveAppBar } from './ResponsiveAppBar';
