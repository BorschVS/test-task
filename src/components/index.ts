export { default as App } from './App';
export { default as Layout } from './Layout';
export { default as CategoryFilter } from './CategoryFilter';
export { default as FlightCard } from './FlightCard';
export { default as TransferFilter } from './TransferFilter';
export { default as ResponsiveAppBar } from './ResponsiveAppBar';
export { default as ContactModal } from './ContactModal';
export { default as Form } from './Form';
export { default as AvatarImage } from './AvatarImage';
export { default as BackgroundImage } from './BackgroundImage';
