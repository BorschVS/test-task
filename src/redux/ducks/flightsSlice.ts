import { createSlice, PayloadAction } from '@reduxjs/toolkit';

import { FlightData, FlightsState } from 'types/interfaces';
import { FlightCategory } from 'types/types';

import { FAST_VALUE, CHEAP_VALUE } from '../../constants';

export const GET_FLIGHTS = 'GET_FLIGHTS';
export const SET_FLIGHTS = 'SET_FLIGHTS';
export const SET_FLIGHTS_LOADING = 'SET_FLIGHTS_LOADING';
export const SET_FLIGHTS_ERROR = 'SET_FLIGHTS_ERROR';
export const SET_STOPS_FILTER = 'SET_STOPS_FILTER';
export const SET_ALL_STOPS = 'SET_ALL_STOPS';
export const SET_STOPS_FILTER_STATUS = 'SET_STOPS_FILTER_STATUS';
export const SET_CATEGORY_FILTER = 'SET_CATEGORY_FILTER';

const initialState: FlightsState = {
  availableFlights: [],
  filteredFlights: [],
  stopsFilter: [],
  stopsFilterStatus: false,
  flightCategoryFilter: CHEAP_VALUE,
  loading: false,
  error: null,
};

const flightsSlice = createSlice({
  name: 'flights',
  initialState,
  reducers: {
    setFlights(state, action: PayloadAction<FlightData[]>) {
      state.availableFlights = action.payload
        .slice()
        .sort((prev, next) => prev.price - next.price);
    },
    setFlightsLoading(state, action: PayloadAction<boolean>) {
      state.loading = action.payload;
    },
    setFlightsError(state, action: PayloadAction<Error | string>) {
      state.error = action.payload;
      state.loading = false;
    },
    setCategoryFilter(state, action: PayloadAction<FlightCategory>) {
      if (action.payload === CHEAP_VALUE) {
        state.flightCategoryFilter = action.payload;
        state.filteredFlights = state.filteredFlights
          .slice()
          .sort((prev, next) => prev.price - next.price);
        state.availableFlights = state.availableFlights
          .slice()
          .sort((prev, next) => prev.price - next.price);
      }
      if (action.payload === FAST_VALUE) {
        state.flightCategoryFilter = action.payload;
        state.filteredFlights = state.filteredFlights
          .slice()
          .sort(
            (prev, next) =>
              prev.segments[0].duration - next.segments[0].duration
          );
        state.availableFlights = state.availableFlights
          .slice()
          .sort(
            (prev, next) =>
              prev.segments[0].duration - next.segments[0].duration
          );
      }
    },
    setStopsFilter(state, action: PayloadAction<number[]>) {
      state.stopsFilter = action.payload;
      state.filteredFlights = state.availableFlights.filter((flight) =>
        flight.segments.some((segment) =>
          action.payload.includes(segment.stops.length)
        )
      );
      state.stopsFilterStatus = false;
    },
    setAllStops(state) {
      state.stopsFilter = [0, 1, 2, 3];
      state.filteredFlights = state.availableFlights;
      state.stopsFilterStatus = true;
    },
    setStopsFilterStatus(state) {
      state.stopsFilterStatus = state.stopsFilter.length > 0;
    },
  },
});

export const {
  setFlights,
  setFlightsLoading,
  setFlightsError,
  setCategoryFilter,
  setStopsFilter,
  setAllStops,
  setStopsFilterStatus,
} = flightsSlice.actions;

export default flightsSlice.reducer;
