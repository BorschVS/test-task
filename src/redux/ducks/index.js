export { default as flights } from './flights';
export { default as form } from './form';
export { default as modal } from './modal';
export { default as images } from './images';