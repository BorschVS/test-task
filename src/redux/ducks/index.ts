export { flightsReducer } from './flights';
export { formReducer } from './form';
export { modalReducer } from './modal';
export { imagesReducer } from './images';
