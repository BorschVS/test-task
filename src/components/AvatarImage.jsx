import { useEffect } from 'react';
import { useDispatch } from 'react-redux';

import { Avatar } from '@mui/material';

import { getImageRequest } from '../redux/ducks/images';

const AvatarImage = ({
  imageName,
  alt,
  component = 'div',
  width = 40,
  height = 40,
  margin = '0',
}) => {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(getImageRequest(imageName));
  }, [dispatch, imageName]);

  return (
    <span>
      <Avatar
        component={component}
        src={`/images/${imageName}`}
        alt={alt}
        sx={{
          display: 'block',
          margin,
          width,
          height,
        }}
      />
    </span>
  );
};

export default AvatarImage;
