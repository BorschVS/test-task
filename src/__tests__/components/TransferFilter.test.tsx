import { screen, fireEvent, render } from '@testing-library/react';
import { useDispatch, useSelector } from 'react-redux';

import { TransferFilter } from 'components';

import { setAllStops, setStopsFilter, setStopsFilterStatus } from '../../redux/ducks/flightsSlice';

jest.mock('react-redux', () => ({
    useDispatch: jest.fn(),
    useSelector: jest.fn(),
}));

jest.mock('../../redux/ducks/flightsSlice', () => ({
    setAllStops: jest.fn(),
    setStopsFilter: jest.fn(),
    setStopsFilterStatus: jest.fn(),
}));

describe('TransferFilter component', () => {
    const mockDispatch = jest.fn();

    beforeEach(() => {
        (useSelector as unknown as jest.Mock).mockReturnValueOnce([0]);
        (useDispatch as unknown as jest.Mock).mockReturnValue(mockDispatch);
    });

    afterEach(() => {
        jest.clearAllMocks();
    })

    test('Should render TransferFilter correctly', () => {
        render(<TransferFilter />);

        const header = screen.getByText('Количество пересадок');
        expect(header).toBeInTheDocument();
    });

    test('Should dispatch setAllStops when "Все" checkbox is checked', () => {
        render(<TransferFilter />);

        const checkboxAll = screen.getByLabelText('Все');
        fireEvent.click(checkboxAll);

        expect(mockDispatch).toHaveBeenCalledWith(setAllStops());
    });

    test('Should dispatch setStopsFilter with empty array when "Все" checkbox is unchecked', () => {
        render(<TransferFilter />);

        const checkboxAll = screen.getByLabelText('Все');
        fireEvent.click(checkboxAll);

        expect(mockDispatch).toHaveBeenCalledWith(setStopsFilter([]));
    });

    test('Should dispatch setStopsFilterStatus after changing the filter', () => {
        render(<TransferFilter />);

        const checkbox = screen.getByLabelText('Без пересадок');
        fireEvent.click(checkbox);

        expect(mockDispatch).toHaveBeenCalledWith(setStopsFilterStatus());
    });

    test('Should selecting "Без пересадок" updates the filter', () => {
        render(<TransferFilter />);

        const checkbox = screen.getByLabelText('Без пересадок');
        fireEvent.click(checkbox);

        expect(mockDispatch).toHaveBeenCalledWith(setStopsFilter([0]));
    });

    test('Should selecting "1 пересадка" updates the filter', () => {
        render(<TransferFilter />);

        const checkbox = screen.getByLabelText('1 пересадка');
        fireEvent.click(checkbox);

        expect(mockDispatch).toHaveBeenCalledWith(setStopsFilter([1]));
    });

    test('Should selecting multiple options updates the filter accordingly', () => {
        render(<TransferFilter />);

        const checkboxOneTransfer = screen.getByLabelText('1 пересадка');
        const checkboxTwoTransfers = screen.getByLabelText('2 пересадки');

        fireEvent.click(checkboxOneTransfer);
        fireEvent.click(checkboxTwoTransfers);

        expect(mockDispatch).toHaveBeenCalledWith(setStopsFilter([1, 2]));
    });

    test('Should check "all" checkbox if all checkboxes is checked', () => {
        render(<TransferFilter />);

        const checkboxWhithoutTransfers = screen.getByLabelText('Без пересадок');
        const checkboxOneTransfer = screen.getByLabelText('1 пересадка');
        const checkboxTwoTransfers = screen.getByLabelText('2 пересадки');
        const checkboxThreeTransfers = screen.getByLabelText('3 пересадки');

        fireEvent.click(checkboxWhithoutTransfers);
        fireEvent.click(checkboxOneTransfer);
        fireEvent.click(checkboxTwoTransfers);
        fireEvent.click(checkboxThreeTransfers);

        expect(mockDispatch).toHaveBeenCalledWith(setStopsFilter([0, 1, 2, 3]));
    });

    test('Should match TransferFilter snapshot', () => {
        const { asFragment } = render(
            <TransferFilter />
        );

        expect(asFragment()).toMatchSnapshot();
    });
});