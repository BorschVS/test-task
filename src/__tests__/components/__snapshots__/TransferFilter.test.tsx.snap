// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransferFilter component Should match FlightCard snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-gnpbhQ bElNHe"
  >
    <h3
      class="MuiTypography-root MuiTypography-h3 css-7a3q38-MuiTypography-root"
    >
      Количество пересадок
    </h3>
    <div
      class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
    >
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd sc-la-DxNn kfTsrL css-j204z7-MuiFormControlLabel-root"
      >
        <span
          class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            type="checkbox"
            value="all"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium sc-iCZwEW gLUodm css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxOutlineBlankOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5zm0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          Все
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd sc-la-DxNn kfTsrL css-j204z7-MuiFormControlLabel-root"
      >
        <span
          class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium Mui-checked MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            checked=""
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            type="checkbox"
            value="0"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium sc-knesRu ciMUDf css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2m0 16H5V5h14zM17.99 9l-1.41-1.42-6.59 6.59-2.58-2.57-1.42 1.41 4 3.99z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          Без пересадок
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd sc-la-DxNn kfTsrL css-j204z7-MuiFormControlLabel-root"
      >
        <span
          class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            type="checkbox"
            value="1"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium sc-iCZwEW gLUodm css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxOutlineBlankOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5zm0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          1 пересадка
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd sc-la-DxNn kfTsrL css-j204z7-MuiFormControlLabel-root"
      >
        <span
          class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            type="checkbox"
            value="2"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium sc-iCZwEW gLUodm css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxOutlineBlankOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5zm0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          2 пересадки
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd sc-la-DxNn kfTsrL css-j204z7-MuiFormControlLabel-root"
      >
        <span
          class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            type="checkbox"
            value="3"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium sc-iCZwEW gLUodm css-i4bv87-MuiSvgIcon-root"
            data-testid="CheckBoxOutlineBlankOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5zm0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          3 пересадки
        </span>
      </label>
    </div>
  </div>
</DocumentFragment>
`;
